src/routes/tutorial/[slug]/adapter.js

```diff
@@ -1,77 +1,88 @@
+import { writable } from 'svelte/store';
+import { browser } from '$app/environment';
+
+export const progress = writable({
+	value: 0,
+	text: 'initialising'
+});
+
+/** @type {import('svelte/store').Writable<string | null>} */
+export const base = writable(null);
+
+/** @type {import('svelte/store').Writable<Error | null>} */
+export const error = writable(null);
+
+let ready = new Promise(() => {});
+
+if (browser) {
+	ready = new Promise(async (fulfil, reject) => {
+		try {
+			const module = await import('$lib/client/adapters/webcontainer/index.js');
+			const adapter = await module.create((value, text) => {
+				progress.set({ value, text });
+			});
+
+			base.set(adapter.base);
+			publish('reload');
+
+			fulfil(adapter);
+		} catch (error) {
+			reject(error);
+		}
+	})
+}
+
+/** @typedef {'reload'} EventName */
+
+/** @type {Map<EventName, Set<() => void>>} */
+let subscriptions = new Map([['reload', new Set()]]);
+
 /**
- * @param {import('$lib/types').Stub[]} initial_stubs
- * @param {(progress: number, status: string) => void} callback
- * @returns {import('$lib/types').Adapter}
+ * 
+ * @param {EventName} event 
+ * @param {() => void} callback 
  */
-export function create_adapter(initial_stubs, callback) {
-	/**
-	 * @typedef {{ type: 'reset'; stubs: import('$lib/types').Stub[]; } | { type: 'update'; stubs: import('$lib/types').FileStub[]; }} State
-	 */
-
-	/** @type {State | undefined} */
-	let state;
-	/** @type {Promise<import('$lib/types').AdapterInternal>} */
-	let adapter_promise;
-	let adapter_base = '';
-
-	async function init() {
-		const module = await import('$lib/client/adapters/webcontainer/index.js');
-		adapter_promise = module.create(initial_stubs, callback);
-		adapter_base = (await adapter_promise).base;
-	}
+export function subscribe(event, callback) {
+	subscriptions.get(event)?.add(callback);
 
-	// Keep track of what's currently running, and what's next
-	/** @type {Promise<boolean>} */
-	let current;
-	let token = {};
-	async function next() {
```

Comment:
The simplification inside `Output.svelte` is great, but I think we need to keep this queue implementation. The new code is - I think - prone to race conditions when typing fast, toggling between solution and initial state fast etc.

Reply:
I haven't observed that to be the case, but if it is then I think we want to implement the queuing logic in `webcontainer/index.js`, where there's already an `await running` (which appears to only be used in the `reset` case) — having multiple layers was part of the reason the code was hard to work with. As far as `adapter.js` goes it should be fire-and-forget, I think